Framework/Algorithms/src/MultipleScatteringCorrection.cpp

```diff
@@ -349,73 +755,207 @@ void MultipleScatteringCorrection::calculateL2Ds(const MultipleScatteringCorrect
+                                               const std::vector<double> &L2Ds,                          //
+                                               const int64_t startIndex, const int64_t endIndex) const {
   if (endIndex - startIndex > MAX_INTEGRATION_LENGTH) {
-    size_t middle = findMiddle(startIndex, endIndex);
+    int64_t middle = findMiddle(startIndex, endIndex);
 
     // recursive to process upper and lower part
-    pairWiseSum(A1, A2, linearCoefAbs, distGraber, L2Ds, L12s, startIndex, middle);
-    pairWiseSum(A1, A2, linearCoefAbs, distGraber, L2Ds, L12s, middle, endIndex);
+    pairWiseSum(A1, A2, linearCoefAbs, distGraber, LS1s, L12s, L2Ds, startIndex, middle);
+    pairWiseSum(A1, A2, linearCoefAbs, distGraber, LS1s, L12s, L2Ds, middle, endIndex);
   } else {
     // perform the integration
-    const auto &Ls1s = distGraber.m_LS1;
     const auto &elementVolumes = distGraber.m_elementVolumes;
     const auto nElements = distGraber.m_numVolumeElements;
-    for (size_t i = startIndex; i < endIndex; ++i) {
+    for (int64_t i = startIndex; i < endIndex; ++i) {
       // compute A1
-      double exponent = (Ls1s[i] + L2Ds[i]) * linearCoefAbs;
+      double exponent = (LS1s[i] + L2Ds[i]) * linearCoefAbs;
       A1 += exp(exponent) * elementVolumes[i];
       // compute A2
       double a2 = 0.0;
-      for (size_t j = 0; j < nElements; ++j) {
+      for (int64_t j = 0; j < int64_t(nElements); ++j) {
         if (i == j) {
           // skip self (second order scattering must happen in a different element)
           continue;
         }
         // L12 is a pre-computed vector, therefore we can use the index directly
         size_t idx_l12 = i < j ? calcLinearIdxFromUpperTriangular(nElements, i, j)
                                : calcLinearIdxFromUpperTriangular(nElements, j, i);
+        if (L12s[idx_l12] == 0) {
+          continue;
+        }
         // compute a2 component
-        exponent = (Ls1s[j] + L12s[idx_l12] + L2Ds[i]) * linearCoefAbs;
+        exponent = (LS1s[i] + L12s[idx_l12] + L2Ds[j]) * linearCoefAbs;
         a2 += exp(exponent) * elementVolumes[j] / (L12s[idx_l12] * L12s[idx_l12]);
       }
       A2 += a2 * elementVolumes[i];
     }
   }
 }
 
+/**
+ * @brief Integrate for A1 and A2 using pair-wise summation for multi component case
+ *
+ * @param A1
+ * @param A2
+ * @param linearCoefAbsContainer
+ * @param linearCoefAbsSample
+ * @param numVolumeElementsContainer
+ * @param numVolumeElementsTotal
+ * @param totScatterCoefContainer
+ * @param totScatterCoefSample
+ * @param elementVolumes
+ * @param LS1sContainer
+ * @param LS1sSample
+ * @param L12sContainer
+ * @param L12sSample
+ * @param L2DsContainer
+ * @param L2DsSample
+ * @param startIndex
+ * @param endIndex
+ */
+void MultipleScatteringCorrection::pairWiseSum(
+    double &A1, double &A2,                                                          //
```

Comment:
empty trailing comments

Reply:
As mentioned above, these comments are used to separate input args input different categories so that other developers have a better chance understanding the input structure.